### ISSUE TITLE
how to use tensorrt python api to implement attention+mask?

### ISSUE THREAD

**Issue by jike-algorithm-zhangxiao** [NVIDIA/TensorRT#2510]
I met a transformer which attention implementation like this:

```

import torch
import torch.nn as nn
import numpy as np
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.query = nn.Linear(1024, 1024)
        self.key = nn.Linear(1024, 1024)
        self.value = nn.Linear(1024, 1024)

    def forward(self, x, mask):
        q = self.query(x)
        k = self.key(x)
        v = self.value(x)

        q = q.view(1, 3, 16, 64).permute(0, 2, 1, 3) * (64 ** -0.5)
        k = k.view(1, 3, 16, 64).permute(0, 2, 3, 1)
        v = v.view(1, 3, 16, 64).permute(0, 2, 1, 3)

        qk = q @ k
        qk = qk + mask
        w = F.softmax(qk.float(), dim=-1)
        return (w @ v).permute(0, 2, 1, 3).reshape(1, 3, 1024)

if __name__ == '__main__':
    input = torch.randn((1, 3, 1024))
    mask = torch.empty(3, 3).fill_(-np.inf).triu_(1)
    model = Model()
    x = model(input, mask)
    print(x.shape)


```

How can i use python api to accomplish this?
I am know how to use CustomQKVToContextPluginDynamic, but the triu mask add troubled me long time.
Could you give me some advice?

**Comment by zerollzeng** [comment]
I think it's possible, please refer to https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/ @pranavm-nvidia to double confirm.

**Comment by zerollzeng** [comment]
also refer to https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Layers.html

**Comment by jike-algorithm-zhangxiao** [comment]
I wanna make this attention forward as a plugin which fused the process about cal the context.
without triu mask, CustomQKVToContextPluginDynamic can do this.
may this plugin can support extra add mask operation? if support, how to use it?

Also i found if i use api to construct this network, such as 
```
    q = network.add_shuffle(out(q))
    q.reshape_dims = (1, -1, 12, 64)
    q.second_transpose = (0, 2, 1, 3)  # 1 12 -1 64
    k = network.add_shuffle(out(k))
    k.reshape_dims = (1, -1, 12, 64)
    k.second_transpose = (0, 2, 1, 3)  # 1 12 -1 64
    v = network.add_shuffle(out(v))
    v.reshape_dims = (1, -1, 12, 64)
    v.second_transpose = (0, 2, 1, 3)  # 1 12 -1 64
    qk_score = network.add_matrix_multiply(out(q), trt.MatrixOperation.NONE, out(k), trt.MatrixOperation.TRANSPOSE)  # 1 12 -1 -1
    qk_masks = network.add_elementwise(out(qk_score), input_masks, trt.ElementWiseOperation.SUM)
    qk_score = network.add_softmax(out(qk_masks))
    qk_score.axes = 1 << 3  # 1 12 -1 -1
    context = network.add_matrix_multiply(out(qk_score), trt.MatrixOperation.NONE, out(v), trt.MatrixOperation.NONE)  # 1 12 -1 64
    context = network.add_shuffle(out(context))
    context.first_transpose = (0, 2, 1, 3)
    context.reshape_dims = (1, -1, 768) # 1 -1 768
```
when i build the tensorrt, the verbose show me it will use a ForeignNode to impelment this.
Is tensorrt automatic supported to fuse attention-like process?
and what's the difference between foreignnode and plugins such as  CustomQKVToContextPluginDynamic?

**Comment by zerollzeng** [comment]
@ttyio Any suggestion here?

**Comment by ttyio** [comment]
You could see it is multiple cuda kernel in today implementation if you run the foreignNode under nsight. The long term goal is to generate single fused kernel like `CustomQKVToContextPluginDynamic`.

**Comment by jike-algorithm-zhangxiao** [comment]
thanks !
